src/hotspot/share/prims/jvmtiImpl.cpp

```diff
@@ -644,6 +644,18 @@ void VM_BaseGetOrSetLocal::doit() {
     return;
   }
   if (_set) {
+    if (fr.is_heap_frame()) { // we want this check after the check for JVMTI_ERROR_INVALID_SLOT
+      assert(Continuation::is_frame_in_continuation(_jvf->thread(), fr), "sanity check");
+      // If the topmost frame is a heap frame, then it hasn't been thawed. This can happen
+      // if we are executing at a return barrier safepoint. The callee frame has been popped,
+      // but the caller frame has not been thawed. We can't support a JVMTI SetLocal in the callee
+      // frame at this point, because we aren't truly in the callee yet.
+      // fr.is_heap_frame() is impossible if a continuation is at a single step or breakpoint.
+      // In such cases the top frames can't be frozen because of an agent callback frame.
```

Comment:
"frames" should be "frame".  I don't understand the agent callback frame reference.

Reply:
Removed this sentence as unneeded.